drake/solvers/MobyLCP.cpp

```diff
@@ -0,0 +1,1278 @@
+// Adapted with permission from code by Evan Drumwright
+// (https://github.com/edrumwri).
+
+#include "MobyLCP.h"
+
+#include <Eigen/LU>
+#include <Eigen/SparseCore>
+#include <Eigen/SparseLU>
+
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include "Optimization.h"
+
+namespace {
+// TODO sammy these helpers probably have some faster, more Eigen way
+// to do this than I've come up with.  This implementation, anyway, is
+// definitely slow.  It's an attempt at rewriting some Ravelin
+// functionality.  Also, the only reason that the out paramater is a
+// MatrixXd is so that this function can resize it.  Runtime
+// performance may be improved by allowing the caller to create a
+// fixed-size matrix and pass that in instead.
```

Comment:
I don't see a faster way actually.

The fixed size matrix optimization could make things faster, but I don't see any way to know the (maximum) dimensions at compile time, as currently not even the maximum dimensions of `in` are known at compile time in usages.

Unless you know that `rows` and `cols` are contiguous (in which case you can use Eigen's `Block`), or more generally, there's some kind of fixed [stride](http://eigen.tuxfamily.org/dox/classEigen_1_1Stride.html), I think you're forced to make copies of the data.